internal/access/grant.go

```diff
@@ -18,33 +22,49 @@ func GetGrant(c *gin.Context, id uid.ID) (*models.Grant, error) {
 }
 
 func ListGrants(c *gin.Context, subject uid.PolymorphicID, resource string, privilege string) ([]models.Grant, error) {
-	db, err := RequireInfraRole(c, models.InfraAdminRole, models.InfraViewRole, models.InfraConnectorRole)
-	if err != nil {
-		return nil, err
+	selectors := []data.SelectorFunc{
+		data.ByOptionalResource(resource),
+		data.ByOptionalPrivilege(privilege),
 	}
 
-	return data.ListGrants(db,
-		data.ByOptionalSubject(subject),
-		data.ByOptionalResource(resource),
-		data.ByOptionalPrivilege(privilege))
-}
+	db, err := RequireInfraRole(c, models.InfraAdminRole, models.InfraViewRole, models.InfraConnectorRole)
+	if err == nil {
+		selectors = append(selectors, data.ByOptionalSubject(subject))
+		return data.ListGrants(db, selectors...)
+	}
 
-func ListIdentityGrants(c *gin.Context, identityID uid.ID) ([]models.Grant, error) {
-	db, err := hasAuthorization(c, identityID, isIdentitySelf, models.InfraAdminRole, models.InfraViewRole)
-	if err != nil {
-		return nil, err
+	if errors.Is(err, internal.ErrForbidden) {
+		// Allow an authenticated identity to view their own grants
+		db := getDB(c)
+		subjectID, _ := subject.ID()
+		identity := AuthenticatedIdentity(c)
+		switch {
+		case identity == nil:
+			return nil, err
+		case subject.IsIdentity() && identity.ID == subjectID:
+			selectors = append(selectors, data.ByOptionalSubject(subject))
+			return data.ListGrants(db, selectors...)
+		case subject.IsGroup() && userInGroup(db, identity.ID, subjectID):
+			selectors = append(selectors, data.ByOptionalSubject(subject))
+			return data.ListGrants(db, selectors...)
+		}
 	}
 
-	return data.ListIdentityGrants(db, identityID)
+	return nil, err
 }
 
-func ListGroupGrants(c *gin.Context, groupID uid.ID) ([]models.Grant, error) {
-	db, err := hasAuthorization(c, groupID, isUserInGroup, models.InfraAdminRole, models.InfraViewRole)
+func userInGroup(db *gorm.DB, authnUserID uid.ID, groupID uid.ID) bool {
```

Comment:
I'd like to see something that can be passed to `hasAuthorization` if possible, this seems like it could be a common pattern. If that requires a lot of changes, then this is ok as is.

Reply:
I think that would be nice. I would like to do that as well, but my thinking was that maybe it would be better as a follow up. This is still a contentious change, so I didn't want to go with a big refactor up front.

I'll explore what that might look like now.

Reply:
I'm good with this suggestion.